src/send_future.rs

```diff
@@ -0,0 +1,180 @@
+use crate::manager::AddressMessage;
+use crate::receiver::Receiver;
+use crate::Disconnected;
+use flume::r#async::SendFut;
+use futures_core::future::BoxFuture;
+use futures_util::FutureExt;
+use std::future::Future;
+use std::marker::PhantomData;
+use std::mem;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// A [`Future`] that represents the state of sending a message to an actor.
+///
+/// By default, a `SendFuture` will also wait until the handler has finished executing and resolve to the return value (see [`Handler::Return`](crate::Handler::Return)).
+/// This behaviour can be changed by calling [`recv_async`](SendFuture::recv_async).
+///
+/// When toggled to receive the return value asynchronously, this future will resolve once the message is successfully queued into the actor's mailbox. If the actors mailbox is bounded, this future will yield `Pending` until there is space in the mailbox. This allows an actor to exercise backpressure on its users.
+#[must_use = "Futures do nothing unless polled"]
+pub struct SendFuture<R, F, TRecvSyncMarker> {
+    inner: SendFutureInner<R, F>,
+    phantom: PhantomData<TRecvSyncMarker>,
+}
+
+/// Marker type for receiving the return value of a [`Handler`](crate::Handler) **synchronously**.
+pub enum ReceiveSync {}
+
+/// Marker type for receiving the return value of a [`Handler`](crate::Handler) **asynchronously**.
+pub enum ReceiveAsync {}
+
+enum SendFutureInner<R, F> {
+    Disconnected,
+    Sending(F),
+    Receiving(Receiver<R>),
+    Done,
+}
+
+impl<R, F> SendFuture<R, F, ReceiveSync> {
+    pub(crate) fn disconnected() -> Self {
+        Self {
+            inner: SendFutureInner::Disconnected,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Toggle this future to receive the return value asynchronously.
+    ///
+    /// Calling this function will change the [`Output`](Future::Output) of this [`Future`] from [`Handler::Return`](crate::Handler::Return) to [`Receiver<Handler::Return>`](Receiver<crate::Handler::Return>).
+    ///
+    /// A [`Receiver`] is a [`Future`] itself that can be used to await the execution of the [`Handler`](crate::Handler) separately from the queuing of the message.
+    pub fn recv_async(self) -> SendFuture<R, F, ReceiveAsync> {
+        SendFuture {
+            inner: self.inner,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<R> SendFuture<R, BoxFuture<'static, Receiver<R>>, ReceiveSync> {
+    pub(crate) fn sending_boxed<F>(send_fut: F) -> Self
```

Comment:
Just to check - this is used for message channels, whilst sending_named is to specific actor types?

Reply:
Correct!